ang/civicase/ActivityCard.js

```diff
@@ -1,56 +1,103 @@
-(function(angular, $, _) {
+(function (angular, $, _) {
+  var module = angular.module('civicase');
 
-  function activityCard($scope, getActivityFeedUrl, dialogService, templateExists) {
+  module.directive('caseActivityCard', function () {
+    return {
+      restrict: 'A',
+      templateUrl: '~/civicase/ActivityCard.html',
+      controller: caseActivityCardController,
+      scope: {
+        activity: '=caseActivityCard',
+        refresh: '=refreshCallback',
+        editActivityUrl: '=?editActivityUrl'
+      }
+    };
+  });
+
+  function caseActivityCardController ($scope, getActivityFeedUrl, dialogService, templateExists, crmApi) {
     var ts = $scope.ts = CRM.ts('civicase');
-    $scope.CRM = CRM;
     $scope.activityFeedUrl = getActivityFeedUrl;
+    $scope.CRM = CRM;
     $scope.templateExists = templateExists;
 
-    $scope.isActivityEditable = function(activity) {
+    // console.log($scope.activity.category);
+
+    $scope.isActivityEditable = function (activity) {
       var type = CRM.civicase.activityTypes[activity.activity_type_id].name;
+
       return (type !== 'Email' && type !== 'Print PDF Letter') && $scope.editActivityUrl;
     };
 
-    $scope.markCompleted = function(act) {
-      $scope.refresh([['Activity', 'create', {id: act.id, status_id: act.is_completed ? 'Scheduled' : 'Completed'}]]);
+    /**
+     * Mark an activity as complete
+     *
+     * @param {object} activity
+     */
+    $scope.markCompleted = function (activity) {
+      return crmApi([['Activity', 'create', {id: activity.id, status_id: activity.is_completed ? 'Scheduled' : 'Completed'}]])
+        .then(function (data) {
+          if (!data[0].is_error) {
+            activity.is_completed = !activity.is_completed;
+          }
+        });
     };
 
-    $scope.star = function(act) {
-      act.is_star = act.is_star === '1' ? '0' : '1';
+    /**
+     * Star/Favourite an activity
+     *
+     * @param {object} $event
+     * @param {object} activity
+     */
+    $scope.star = function ($event, activity) {
```

Comment:
Shouldn't the name be more descriptive. Something like `$scope.starIt` or `$scope.markItAsFav` etc. ?